allennlp/data/dataset_readers/reading_comprehension/dqa.py

```diff
@@ -0,0 +1,127 @@
+
+
+@DatasetReader.register("dqa")
+class DQAReader(DatasetReader):
+    """
+    Reads a JSON-formatted QUAC file and returns a ``Dataset`` where the ``Instances`` have four
+    fields: ``question``, a ``TextField``, ``passage``, another ``TextField``, and ``span_start``
+    and ``span_end``, both ``IndexFields`` into the ``passage`` ``TextField``.  We also add a
+    ``MetadataField`` that stores the instance's ID, the original passage text, gold answer strings,
+    and token offsets into the original passage, accessible as ``metadata['id']``,
+    ``metadata['original_passage']``, ``metadata['answer_text_lists'] and ``metadata['token_offsets']``.  
+
+    Parameters
+    ----------
+    tokenizer : ``Tokenizer``, optional (default=``WordTokenizer()``)
+        We use this ``Tokenizer`` for both the question and the passage.  See :class:`Tokenizer`.
+        Default is ```WordTokenizer()``.
+    token_indexers : ``Dict[str, TokenIndexer]``, optional
+        We similarly use this for both the question and the passage.  See :class:`TokenIndexer`.
+        Default is ``{"tokens": SingleIdTokenIndexer()}``.
+    prev_a : Integer, optional 
+        How many previous questions to consider.
+    """
+    def __init__(self,
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 lazy: bool = False,
+                 prev_a: int = 0) -> None:
+        super().__init__(lazy)
+        self._tokenizer = tokenizer or WordTokenizer()
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+        self._prev_a = prev_a
+
+    @overrides
+    def _read(self, file_path: str):
+        # if `file_path` is a URL, redirect to the cache
+        file_path = cached_path(file_path)
+        logger.info("Reading file at %s", file_path)
+        with open(file_path) as dataset_file:
+            dataset_json = json.load(dataset_file)
+            dataset = dataset_json['data']
+        logger.info("Reading the dataset")
+        for article in dataset:
+            for paragraph_json in article['paragraphs']:
+                paragraph = paragraph_json["context"]
+                tokenized_paragraph = self._tokenizer.tokenize(paragraph)
+                qas = paragraph_json['qas']
+                metadata = {}
+                metadata["instance_id"] = [qa['id'] for qa in qas]
+                question_text_list = [qa["question"].strip().replace("\n", "") for qa in qas]
+                answer_texts_list = [[answer['text'] for answer in qa['answers']] for qa in qas]
+                metadata["question"] = question_text_list
+                metadata['answer_texts_list'] = answer_texts_list
+                span_starts_list = [[answer['answer_start'] for answer in qa['answers']] for qa in qas]
+                span_ends_list = []
+                for st_list, an_list in zip(span_starts_list, answer_texts_list):
```

Comment:
same here - I have to really study the code to know what st_list is, whereas if you'd called it `answer_starts`, I can immediately parse.